src/java/org/apache/cassandra/utils/bytecomparable/ByteSourceInverse.java

```diff
@@ -0,0 +1,436 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.utils.bytecomparable;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+import org.apache.cassandra.db.marshal.ValueAccessor;
+
+/**
+ * Contains inverse transformation utilities for {@link ByteSource}s.
+ *
+ * See ByteComparable.md for details about the encoding scheme.
+ */
+public final class ByteSourceInverse
+{
+    private static final int INITIAL_BUFFER_CAPACITY = 32;
+
+    /**
+     * Get the given number of bytes and produce a long from them, effectively treating the bytes as a big-endian
+     * unsigned encoding of the number.
+     */
+    public static long getUnsignedFixedLengthAsLong(ByteSource byteSource, int length)
+    {
+        if (byteSource == null)
+            throw new IllegalArgumentException("Unexpected null ByteSource");
+        if (length < 1 || length > 8)
+            throw new IllegalArgumentException("Between 1 and 8 bytes can be read at a time");
+
+        long result = 0;
+        for (int i = 0; i < length; ++i)
+        {
+            int data = byteSource.next();
+            if (data == ByteSource.END_OF_STREAM)
+                throw new IllegalArgumentException(
+                        String.format("Unexpected end of stream reached after %d bytes (expected >= %d)", i, length));
+            assert data >= 0 && data <= 255;
+            result = (result << 8) | data;
+        }
+        return result;
+    }
+
+    /**
+     * Produce the bytes for an encoded signed fixed-length number.
+     * The first byte has its sign bit inverted, and the rest are passed unchanged.
+     */
+    public static <V> V getSignedFixedLength(ValueAccessor<V> accessor, ByteSource byteSource, int length)
+    {
+        if (byteSource == null)
+            throw new IllegalArgumentException("Unexpected null ByteSource");
+        if (length < 1)
+            throw new IllegalArgumentException("At least 1 byte should be read");
+
+        V result = accessor.allocate(length);
+        // The first byte needs to have its sign flipped
+        accessor.putByte(result, 0, (byte) (byteSource.next() ^ 0x80));
```

Comment:
It was `SIGN_MASK` - maybe you could revert to use that constant?

Reply:
Added constants for everything in this file. TBH to me this is harder to read -- I find it easier to understand what bit manipulation does if the constants are given directly.

Reply:
I wouldn't push on that, it was just my personal opinion :)